File: tang/src/main/java/com/microsoft/tang/formats/ParameterParser.java
package com.microsoft.tang.formats;

import java.lang.reflect.Constructor;

import com.microsoft.tang.exceptions.BindException;
import com.microsoft.tang.types.ExternalConstructor;
import com.microsoft.tang.util.MonotonicMap;
import com.microsoft.tang.util.ReflectionUtilities;

public class ParameterParser {
  MonotonicMap<String, Constructor<? extends ExternalConstructor<?>>> parsers = new MonotonicMap<>();

  public void addParser(Class<? extends ExternalConstructor<?>> ec) throws BindException {
    Constructor<? extends ExternalConstructor<?>> c;
    try {
      c = ec.getConstructor(String.class);
    } catch (NoSuchMethodException e) {
      throw new BindException("Constructor "
          + ReflectionUtilities.getFullName(ec) + "(String) does not exist!", e);
    }
    c.setAccessible(true);
    Class<?> tc = ReflectionUtilities.getInterfaceTarget(
        ExternalConstructor.class, ec);
    parsers.put(ReflectionUtilities.getFullName(tc), c);
  }

  public void mergeIn(ParameterParser p) {
    for (String s : p.parsers.keySet()) {
      if (!parsers.containsKey(s)) {
        parsers.put(s, p.parsers.get(s));
      } else {
        if (!parsers.get(s).equals(p.parsers.get(s))) {
          throw new IllegalArgumentException(
              "Conflict detected when merging parameter parsers! To parse " + s
                  + " I have a: " + parsers.get(s)
                  + " the other instance has a: " + p.parsers.get(s));
        }
      }
    }
  }

  public <T> T parse(Class<T> c, String s) {
    Class<?> d = ReflectionUtilities.boxClass(c);
    String name = ReflectionUtilities.getFullName(d);
    return parse(name, s);
  }
  @SuppressWarnings("unchecked")
  public <T> T parse(String name, String value) {
    if (parsers.containsKey(name)) {
      try {
        return (T) (parsers.get(name).newInstance(value).newInstance());
      } catch (ReflectiveOperationException e) {
        throw new IllegalArgumentException("Error invoking constructor for "
            + name, e);
      }
    } else {
      if (name.equals(String.class.getName())) {
        return (T) value;
      }
      if (name.equals(Byte.class.getName())) {
        return (T) (Byte) Byte.parseByte(value);
      }
      if (name.equals(Character.class.getName())) {
        return (T) (Character) value.charAt(0);
      }
      if (name.equals(Short.class.getName())) {
        return (T) (Short) Short.parseShort(value);
      }
      if (name.equals(Integer.class.getName())) {
        return (T) (Integer) Integer.parseInt(value);
      }
      if (name.equals(Long.class.getName())) {
        return (T) (Long) Long.parseLong(value);
      }
      if (name.equals(Float.class.getName())) {
        return (T) (Float) Float.parseFloat(value);
      }
      if (name.equals(Double.class.getName())) {
        return (T) (Double) Double.parseDouble(value);
      }
      if (name.equals(Void.class.getName())) {
        throw new ClassCastException("Can't instantiate void");
      }
      throw new UnsupportedOperationException("Don't know how to parse a " + name);
    }
  }

}

File: tang/src/main/cs/Tang/Implementations/ParameterParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Com.Microsoft.Tang.Interface;
using Com.Microsoft.Tang.Util;

namespace Com.Microsoft.Tang.Implementations
{
    public class ParameterParser
    {
        MonotonicTreeMap<String, ConstructorInfo> parsers = new MonotonicTreeMap<String, ConstructorInfo>();

        public void AddParser(Type ec)
        {
            //Class<?> tc = (Class<?>)ReflectionUtilities.getInterfaceTarget(ExternalConstructor.class, ec);
            //addParser((Class)tc, (Class)ec);
        }

    
        public object Parse(Type c, String s) 
        {
            Type d = ReflectionUtilities.BoxClass(c);
            foreach (Type e in ReflectionUtilities.ClassAndAncestors(d)) 
            {
                string name = e.FullName;
                if (parsers.ContainsKey(name)) 
                {
                    object ret = Parse(name, s);
                    if (c.IsAssignableFrom(ret.GetType())) 
                    {
                        return ret;
                    } 
                    else 
                    {
                        throw new InvalidCastException("Cannot cast from " + ret.GetType() + " to " + c);
                    }
                }
            }
            return Parse(d.Name, s);
        }

        public object Parse(string name, string value) 
        {
            if (parsers.ContainsKey(name)) 
            {
                try 
                {
                    ConstructorInfo c = null;
                    parsers.TryGetValue(name, out c);
                    return ((IExternalConstructor)c.Invoke(new object[] { value })).NewInstance();
                }
                catch (TargetInvocationException e) 
                {
                   throw new ArgumentException("Error invoking constructor for "
                            + name, e);
                }
            } 
            else if (name.Equals(typeof(string).Name)) 
            {
                return (object) value;
            }
            if (name.Equals(typeof(Byte).Name)) 
            {
                return (object) (Byte) Byte.Parse(value);
            }
            if (name.Equals(typeof(Char).Name)) {
                return (object) (Char) value[0];
            }
            //if (name.Equals(typeof(Short).Name)) 
            //{
            //    return (T) (Short) Short.parseShort(value);
            //}
            if (name.Equals(typeof(int).Name))
            {
                return (object) (Int32) Int32.Parse(value);
            }
            //if (name.Equals(Long.class.getName())) {
            //    return (T) (Long) Long.parseLong(value);
            //}
            //if (name.Equals(typeof(Float).Name)) 
            //{
            //    return (object) (Float) Float.parseFloat(value);
            //}
            if (name.Equals(typeof(Double).Name) )
            {
                return (object) (Double) Double.Parse(value);
            }
            if (name.Equals(typeof(Boolean).Name)) 
            {
                return (object) (Boolean) Boolean.Parse(value);
            }
            //if (name.Equals(Void.class.getName())) {
            //    throw new ClassCastException("Can't instantiate void");
            //}
            throw new NotSupportedException("Don't know how to parse a " + name);
       }
    }
}
